### ISSUE TITLE
Infinite loop in `TileCanvasViewGroup.drawTiles`

### ISSUE THREAD

**Issue by p-lr** [moagrius/TileViewLabs#1]
Not an easy one to explain, but sometimes, after a DetailLevel change, and without any user interaction it keeps logging :

> rect=...
> isInViewport? true, isUnderNewTiles? false
> <true> previous tile at ...
> drawing x previous tiles
> there's more work to do, partially opaque tiles were drawn

It appears to me that in the while loop l.253 the boolean `shouldDrawPreviousTile` is always true, given the conditions. The consequence is that the tiles from `mPreviousLevelDrawnTiles` are never removed because `shouldInvalidate` is always true (though i'm not sure why). 
As quick fix, in the previously mentionned while loop, i added a `tile.computeProgress()` :

```
if (shouldDrawnPreviousTile) {
   tile.computeProgress();
   tile.draw(canvas);
   shouldInvalidate = shouldInvalidate || tile.getIsDirty();
} else {
   tilesFromLastDetailLevelIterator.remove();
}
```

After that change, i haven't reproduced the infinite loop yet.

EDIT : i edited my first post because if think it was misleading.

**Comment by p-lr** [comment]
So i did other tests around this issue and i can add those information :
- it's easier to reproduce when zooming in and out between two detail levels, to produce a fade in effect for the new tiles, hence old tiles progressively disappear.
- `shouldInvalidate` is false before entering the while loop. And logging the mProgress attribute for each tile showed values between 0 and 1 exclusively (for example, 0.548), but never updating. It explains why adding `tile.computeProgress()`, which updates mProgress, made the call of `tile.getIsDirty()` return false at some point, thus `shouldInvalidate` stays false.

**Comment by moagrius** [comment]
as always, great feedback Peter, i'll get on these asap

i believe you already understand what's happening really well, but for clarity:
1.  tile.computeProgress updates mProgress with how opaque a tile should be.  this should always increase, as time passes between each call.  if this is not changing, there's a problem
2.  tile.getIsDirty should return false if the tile is not completely opaque, meaning that we need to invalidate and draw again (so it'll get recomputed at a more opaque state, each time)
3.  i had to separate computeProgress from just checking progress during draw, because I need to know if the tile is opaque before the _previous_ tile set gets drawn... if i draw the new tile set, then the old tile set, the old one will be on top...  but i need to know the status of the new set before i draw the old set, since i don't want to bother drawing any old tiles that are going to be completely underneath totally opaque new tiles...

generally speaking, does this look/feel about right to you?  does it seem like this is going to fix the underlying issues that motivated the change?

**Comment by p-lr** [comment]
Generally speaking, i think that transitions between detail levels are now better handled. If i understand correctly, when a detail level change happens, we first attempt to draw tiles from the previous tile set that are :
- in the current viewport (in other words : visible)
- not underneath fully opaque new tiles
  Then, new tiles are drawn. The process is repeated until all tiles from the new set are fully opaque and there is no need to draw old tiles. This is at this moment that the old set is cleared.

I believe this ensures two points :
- smooth transitions when changing detail level
- rendering is consistent because we ensure not to draw old tiles at the end of the drawTiles method

That last point is very important because we can now remove the lock when zooming. This was our main issue : it could cause OOM on big maps - and it still can, as we still give our TileRenderPoolExecutor the list of tiles visible at the current detail level, which can be huge when zooming out from scale 1 to 1/2^25...
Now, programmatic zooms could be a concern if the scale changes too fast. However, we know the destination scale in that case. So we could directly set the current detail level to the destination one, and start render when the scale animation ends. Problem solved :D
What do you think?

Now back to the issue, i think that my suggestion in OP to add `tile.computeProgress()` before drawing an old tile may not be a good idea. In the case of this issue, some old tiles are dirty (not fully opaque), and stay that way. They are visible and not underneath fully opaque new tiles, so illegible for being drawn. But what's the point finishing the old tiles drawing animation if it's to replace it by new tiles? It could make things smoother, idk. To sum up, when an old tile is dirty, i see two options :
- finish the draw by adding `tile.computeProgress()` like suggested in OP
- just remove that tile

Another way to tackle this would be to clear the old tile set when all tiles from the new set are fully opaque. This is what happens already, but only when 'shouldInvalidate` becomes false, which never happens because of some dirty old tiles.

**Comment by moagrius** [comment]
You're exactly correct in all your statements about how things have changed.

> That last point is very important because we can now remove the lock when zooming. This was our main issue : it could cause OOM on big maps - and it still can, as we still give our TileRenderPoolExecutor the list of tiles visible at the current detail level, which can be huge when zooming out from scale 1 to 1/2^25...

True, and actually I'm temporarily preventing any new render requests during scale operations - I'm still not sure how best to handle it, but wanted to tighten things up so I could really see what was happening, visually...

> Now, programmatic zooms could be a concern if the scale changes too fast. However, we know the destination scale in that case. So we could directly set the current detail level to the destination one, and start render when the scale animation ends. Problem solved :D
> What do you think?

That sounds great, and is actually pretty similar to my debug state (where I prevent any requests while zooming).  Do you want to post the specific code changes, or should I just do what I think will achieve what you're describing?

> Now back to the issue, i think that my suggestion in OP to add tile.computeProgress() before drawing an old tile may not be a good idea. In the case of this issue, some old tiles are dirty (not fully opaque), and stay that way. They are visible and not underneath fully opaque new tiles, so illegible for being drawn. But what's the point finishing the old tiles drawing animation if it's to replace it by new tiles? It could make things smoother, idk. To sum up, when an old tile is dirty, i see two options :
> 
> finish the draw by adding tile.computeProgress() like suggested in OP
> just remove that tile
> Another way to tackle this would be to clear the old tile set when all tiles from the new set are fully opaque. This is what happens already, but only when 'shouldInvalidate` becomes false, which never happens because of some dirty old tiles.

Now that I've had a chance to look at the code in relation to what you wrote, it makes perfect sense and I agree with both your diagnosis and fix: just add computeProgress - in fact any time a tile is drawn is should have had computeProgress called just before - which is probably why i missed it - it used to be a single method call, but now (for reasons mentioned yesterday) i had to split it up for timing purposes.  i'm confident adding that line will fix the issue.

This all sounds great - do you think I'm OK to do the above mentioned changes, clean it up and submit for a real PR, or is there anything else that you think should be addressed first?

**Comment by moagrius** [comment]
also just want to confirm you're on branch `complex-draw`

**Comment by p-lr** [comment]
Yes i'm on `complex-draw` branch, and OK for applying the changes, cleanup, etc. About the programmatic zoom,  i can do it later this week.

Some small remarks about the code : 
- `mDecodedTilesInCurrentViewport` is only used in `drawTiles`. And the way it is used makes me think we could avoid creating this temporary list. For example, l.229 of TileCanvasViewGroup, we could just write :

```
for( Tile tile : mTilesInCurrentViewport ) {
      if( tile.getState() == Tile.State.DECODED ) {
          ...
```
- maybe the `cleanup` method could be renamed `keepVisibleTiles`.
  BTW, that cleanup method is called by `beginRenderTask`. So perhaps its other call in `mRenderPostExecuteRunnable` is useless.

**Comment by moagrius** [comment]
> ...and OK for applying the changes, cleanup, etc. About the programmatic zoom, i can do it later this week.

I'll go ahead and cleanup, test, etc this branch and hopefully PR without the zoom stuff then, and we can add that on another branch (but probably before any of this makes it into master - we'll see how things turn out).

> mDecodedTilesInCurrentViewport is only used in drawTiles. And the way it is used makes me think we could avoid creating this temporary list. For example, l.229 of TileCanvasViewGroup, we could just write...

Yeah I went back on forth a lot on this one, and had it both ways.  Since we're already testing decoded state in the first loop, it made sense to me to mark it by placing it in the temporary set, so we don't have to test again, but TBH I have no idea which is more performant - iterating through a larger set and testing a state, or populating a "buffer" set - do you have any metrics?

> maybe the cleanup method could be renamed keepVisibleTiles. 

I'll do that, then alias `cleanup` to `keepVisibleTiles` with a `@deprecation` so we don't break existing builds.

> BTW, that cleanup method is called by beginRenderTask. So perhaps its other call in mRenderPostExecuteRunnable is useless.

You might be right.  I'll put a log around that and test.

Great to have your help Peter.

**Comment by p-lr** [comment]
> Yeah I went back on forth a lot on this one, and had it both ways. Since we're already testing decoded state in the first loop, it made sense to me to mark it by placing it in the temporary set, so we don't have to test again, but TBH I have no idea which is more performant - iterating through a larger set and testing a state, or populating a "buffer" set - do you have any metrics?

No i don't, but it's not very important anyway. It's fine if you let it as is.

> You might be right. I'll put a log around that and test.

Nice.

> Great to have your help Peter.

You welcome!
